Wum live sync automation

ChinthakaSenanayaka opened the pull request:
## Purpose
> This is Jenkins CI automation project for WSO2 support update testing process. 
Here, Jenkins CI is explained with how to run tests against WSO2 products. Review the Groovy and Pipeline scripts.

## Goals
> Auto trigger and build UAT tests in parallel ECS nodes cluster to minimize UAT testing overall time

## Approach
> uatBatchJob batch job -> parallelTriggerJob batch job -> slave product-version jobs on ECS nodes

## User stories
> N/A

## Release note
> Refer to wum-live-sync-automation/README.md

## Documentation
> Refer to wum-live-sync-automation/README.md
chinthakas@wso2.com
## Training
> N/A

## Certification
> N/A

## Marketing
> N/A

## Automation tests
 - Unit tests 
   > N/A
 - Integration tests
   > N/A

## Security checks
   > N/A

## Samples
> N/A

## Related PRs
> N/A

## Migrations (if applicable)
> N/A

## Test environment
> Refer to https://supportbuild-wilkes-upgrade.wso2.com/jenkins/job/uatBatchJob/
 
## Learning
> N/A

Commit:
wum live sync automation after e2e initial testing done

Commit:
wum live sync automation after e2e initial testing done - readme fix

Commit:
wum live sync automation after e2e initial testing done - readme fix

Commit:
wum live sync automation after e2e initial testing done - readme fix

Commit:
wum live sync automation after e2e initial testing done - readme fix

Commit:
wum live sync automation after e2e initial testing done - readme fix

Commit:
wum live sync automation after e2e initial testing done

Commit:
Adding DB connection variable

Commit:
Adding DB connection variable

Commit:
Adding DB connection variable

Commit:
Adding DB connection variable

Manuri commented on wum-live-sync-automation/docker/Dockerfile:
```diff
@@ -0,0 +1,174 @@
+mkdir -p /build/software/maven \
+&& wget -P /build/software/maven https://archive.apache.org/dist/maven/maven-3/3.0.5/binaries/apache-maven-3.0.5-bin.tar.gz \
+&& tar -xvzf /build/software/maven/apache-maven-3.0.5-bin.tar.gz --directory /build/software/maven \
+&& rm /build/software/maven/apache-maven-3.0.5-bin.tar.gz
```
I believe we probably do not need older maven versions for WUM supported products.

ChinthakaSenanayaka replied:
We thought better we keep them in case we need to test for older versions. Having them will not create a problem, but if we did not have them with ECS docker image, time to time we have to upload new docker images.

Manuri commented on wum-live-sync-automation/docker/Dockerfile:
```diff
@@ -0,0 +1,174 @@
+chmod a+x /build/software/java/jdk-6u33-linux-x64.bin \
+&& cd /build/software/java \
+&& echo "\n" | sh /build/software/java/jdk-6u33-linux-x64.bin \
+&& rm /build/software/java/jdk-6u33-linux-x64.bin
```
AFAIK none of the WUM supported products are on JDK 6.

ChinthakaSenanayaka replied:
We thought better we keep them in case we need to test for older versions. Having them will not create a problem, but if we did not have them with ECS docker image, time to time we have to upload new docker images.

ChinthakaSenanayaka reviewed (commented):
Added comments to review comments

Commit:
Merging with upstream
Merge branch 'master' of github.com:wso2/build-automation-artifacts into wum-live-sync-automation

ChinthakaSenanayaka committed:
Merge pull request #1 from wso2/master

Merge base fork

Commit:
Adding updated readme, dockerfile and Jenkins job config editing scripting

Commit:
Merging with Jenkins UAT testing automation changes
Merge branch 'wum-live-sync-automation'

Commit:
uat_batch_job script fixing

Commit:
Adding logic for job name check and WUM updated pack available

kasunbg commented on wum-live-sync-automation/docker/Dockerfile:
```diff
@@ -0,0 +1,248 @@
+
+# ------------------------------------------------------------------------
+
+FROM ubuntu:trusty
```
This Dockerfile should inherit from the WSO2 public docker image instead of bare ubuntu:trusty IMO since this one inherits lot of content from there. I'm trying to see whether we can reduce the duplicate code.. :)

ChinthakaSenanayaka replied:
Dockerfile cannot be extended, docker images should be extended
dockerfile FROM is used to extend from image not file.
To do that we have to make live docker image public, otherwise, this dockerfile will fail. Thus we cant extend docketfile from the live dockerfile or image

ChinthakaSenanayaka replied:
As discussed, we will go with live ECS image local copy and extend it with this Dockerfile. BTW, in readme, this will be mentioned for further reference.

kasunbg commented on wum-live-sync-automation/parallelTriggerJob/Jenkinsfile:
```diff
@@ -0,0 +1,31 @@
+            }
+            resultJson = null
+            
+            parallel branches
```
In the actual deployment, how many parallel branches are we going to have? We've seen that Jenkins UI becomes quite unresponsive when there are loads of parallel jobs.

ChinthakaSenanayaka replied:
Good point, though I did not do a perf test, with ECS most of the load will be distributed to ECS slaves and as of @maheshika 's experience with the current Jenkins master node, this can be handled. I think, we will go with this normal implementation and in case, we see any need to ECS node usage controlling we will act on it.

kasunbg commented on wum-live-sync-automation/parallelTriggerJob/Jenkinsfile:
```diff
@@ -0,0 +1,31 @@
+            
+            parallel branches
+        } catch(Exception e) {
+            println "ERROR: " + e
```
Does this print the exception stack trace as well?

ChinthakaSenanayaka replied:
Added line to print stacktrace.

kasunbg replied:
e.printStackTrace sends the stack trace to stdout, isn't it? In that case, will this error be visible in the jenkins log? We may have to send the error log to `println` command

kasunbg commented on wum-live-sync-automation/uatBatchJob/Jenkinsfile:
```diff
@@ -0,0 +1,176 @@
+import java.nio.file.*
+import groovy.json.*
+
+def boolean runShellCommands(env, wumProdBaseLoc, cmd, wumUpdateFileNameList, wsUpdateFolderName) {
```
It would be good to add a comment mentioning the purpose of this build job and this method.

ChinthakaSenanayaka replied:
Added doc comments

kasunbg commented on wum-live-sync-automation/uatBatchJob/Jenkinsfile:
```diff
@@ -0,0 +1,176 @@
+    }
+}
+
+def void populateJsonData(row, updateListJson, addUpdateDBDetailsOnly, updateListItemNo, updateDBDetailsItemNo, wumUpdateFileNameList, wumReleaseJobNamesList, wsUpdateFolderName, env) {
```
docs are missing..

ChinthakaSenanayaka replied:
Added doc comments

kasunbg commented on wum-live-sync-automation/uatBatchJob/Jenkinsfile:
```diff
@@ -0,0 +1,176 @@
+    def wumPackLoc = "/build/jenkins-home/workspace/wumInstall/wum"
+    def shellRunStatus = runShellCommands(env, wumProdBaseLoc, 
+        "mkdir -p " + env.WORKSPACE + "/" + wsUpdateFolderName + "/ && " +
+        "find " + wumProdBaseLoc + " -name \\wso2*-[0-9].[0-9].[0-9]?*.zip -type f -delete && " +
```
This regex does not match the API-M's version _1.10.0_. Btw, we need to escape the dot char.

ChinthakaSenanayaka replied:
Fixed shell regex for 1.10.0 and escaped dot char

kasunbg commented on wum-live-sync-automation/uatBatchJob/Jenkinsfile:
```diff
@@ -0,0 +1,176 @@
+    
+    println "=============Running Groovy shell commands end==============="
+    
+    Class.forName("com.mysql.jdbc.Driver")
```
Does the MySQL driver you are using support the newer class name `com.mysql.cj.jdbc.Driver`?

ChinthakaSenanayaka replied:
This is updated to new Mysql driver com.mysql.cj.jdbc.Driver. It works fine.

kasunbg commented on wum-live-sync-automation/uatBatchJob/Jenkinsfile:
```diff
@@ -0,0 +1,176 @@
+    System.properties['UPDATE_LIST_JSON'] = new JsonBuilder(updateListJson).toString()
+    
+} catch(Exception e) {
+    println "ERROR in DB querying, " + e;
```
Same as above.

ChinthakaSenanayaka replied:
Added line to print stacktrace.

kasunbg commented on wum-live-sync-automation/parallelTriggerJob/Jenkinsfile:
```diff
@@ -0,0 +1,31 @@
+import groovy.json.JsonSlurper
+
+node {
```
Missing docs.

ChinthakaSenanayaka replied:
Added doc comments

Commit:
Adding code review changes

Commit:
Adding code review changes: added doc commetns

Commit:
Fixing exception print stack trace to Jenkins job consoel

ChinthakaSenanayaka commented:
Merging after code review and approval
